File: src/ios/ShakeMenu.swift
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import UIKit

/// Cordova shake menu stub. Full preview/channel shake UI is planned for a follow-up release.
final class ShakeMenu {
    private let logger: Logger

    init(plugin: CordovaUpdaterPlugin, viewController: UIViewController?, logger: Logger, gesture: String) {
        self.logger = logger
        logger.debug("ShakeMenu: gesture \(gesture) registered (Cordova stub)")
    }

    func usesGesture(_ gesture: String) -> Bool {
        false
    }

    func stop() {
        // no-op
    }
}

extension CordovaUpdaterPlugin {
    func syncShakeMenuGestureRecognizer() {
        // Cordova stub: gesture UI not implemented yet
    }

    func removeShakeMenuGestureRecognizer() {
        // no-op
    }
}

extension UIApplication {
    class func topViewController(_ base: UIViewController? = UIApplication.shared.connectedScenes
        .compactMap { ($0 as? UIWindowScene)?.keyWindow }
        .first?.rootViewController) -> UIViewController? {
        if let nav = base as? UINavigationController {
            return topViewController(nav.visibleViewController)
        }
        if let tab = base as? UITabBarController, let selected = tab.selectedViewController {
            return topViewController(selected)
        }
        if let presented = base?.presentedViewController {
            return topViewController(presented)
        }
        return base
    }
}

File: src/android/app/capgo/cordova/updater/ShakeMenu.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package app.capgo.cordova.updater;

import android.app.Activity;

/**
 * Cordova shake menu stub. Full preview/channel shake UI is planned for a follow-up release.
 */
public class ShakeMenu implements ShakeDetector.Listener, ThreeFingerPinchDetector.Listener {

    private final Logger logger;

    public ShakeMenu(CordovaUpdaterPlugin plugin, Activity activity, Logger logger, String gesture) {
        this.logger = logger;
        logger.debug("ShakeMenu: gesture " + gesture + " registered (Cordova stub)");
    }

    public boolean usesGesture(String gesture) {
        return false;
    }

    public void stop() {
        // no-op
    }

    @Override
    public void onShakeDetected() {
        logger.debug("ShakeMenu: shake detected (Cordova stub)");
    }

    @Override
    public void onThreeFingerPinchDetected() {
        logger.debug("ShakeMenu: three-finger pinch detected (Cordova stub)");
    }
}
